Snowbridge V2 docs (#5902)
paritytech/polkadot-sdk: bridges/snowbridge/docs/v2.md

```
ReserveAssetDeposited (KLT, 80)
PayFees (KLT, 20)
SetAssetClaimer Alice
AliasOrigin Alice
ExchangeAsset give=(KLT, 20) want=(WETH, 1)
InitiateAssetsTransfer asset=(KLT, 60) remoteFee=(WETH, 1) dest=Ethereum
	DepositAsset (KLT, 60) dest=Bob
```

```
*ReserveAssetDeposited (KLT, 80)
*PayFees (KLT, 20)
*SetAssetClaimer Alice
*AliasOrigin Alice
ExchangeAsset give=(KLT, 20) want=(WETH, 1)
InitiateAssetsTransfer asset=(KLT, 60) remoteFee=(WETH, 1) dest=Ethereum
	DepositAsset (KLT, 60) dest=Bob
	Transact Bob.hello()
```

In all cases, $x_1$ should contain the necessary instructions to:

1. Pay fees for local execution using `PaysFees`
2. Obtain WETH for remote delivery fees.

The `SovereignPaidExporter` on AH will charge a small fee to prevent spamming BH with bridge messages. This is necessary since the `ExportMessage` instruction in message $x_2$ will have no execution fee on BH. For a similar reason, we should also impose a minimum relayer reward of at least the existential deposit 0.1 DOT, which acts as a deposit to stop spamming messages with 0 rewards.

## Step 5: BH executes message x2

Message $x_2$ is parsed by the `SnowbridgeMessageExporter` with the following effects:

> - A bridge command $m$ is committed to MMR structure $M$.

Author: Yeah I need to fix that. Its a binary merkle tree.

I've decided not to change to using an MMR for committed messages in V2. The current design works well to keep the onchain code simple (at the expense of more complicated off-chain logic):
1. Binary merkle tree for message leaf hashes and the root commitment.
2. Full message payloads persisted to storage, but are [pruned](https://github.com/paritytech/polkadot-sdk/blob/8614dc0e055d06de4a3774ac1da0a422b33f34e2/bridges/snowbridge/pallets/outbound-queue/src/lib.rs#L247-L253) in the next block

	)
```

The User Agent/UX layer will need to estimate the fee required to be passed into the `send` method. This may be an issue as we cannot Dry-Run something on Polkadot that has not even been submitted on Ethereum yet. We may need to make RPC API to DryRun and get back the xcm that would be submitted to asset hub.

## Step 2: Relayer relays message to Bridge Hub

Exchange rate is eliminated. Users pay remote delivery costs in ETH, and this amount is sent with the message as WETH. The delivery fee can be claimed by the relayer on BH.

The user agent applies a similar dry-running process as with [Step 2: User agent estimates fees](https://www.notion.so/Step-2-User-agent-estimates-fees-113296aaabef8159bcd0e6dd2e64c3d0?pvs=21).

The message is converted from $m_0$ to $x_0$ during message submission on BH. Dry-running submission will return $x_0$ to the relayer so that it can verify it is profitable.

## Step 3: AH receives $x_0$ from BH

Submitting the message $m_0$ will cause the following XCM, $x_0$, to be built on BH and dispatched to AH.

```
WithdrawAsset (KLT, 100)
ReserveAssetDeposited(WETH, 2)
PayFees (WETH, 1)

> SetAssetClaimer Bob            // derived from beneficiary

Reviewer: This is a good question? Should bob be able to claim it directly on asset hub or should bob have to come through KILT to claim the assets?

> SetAssetClaimer Bob            // derived from beneficiary

Reviewer 2: I would say come through KILT. Maybe `Bob` is not a signed address, but a sov-account or pure proxy or some smart contract address and it might only exist on Kilt.

AliasOrigin Ethereum/Alice     // derived from msg.sender
InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
	DepositAsset (KLT, 100) dest=Bob
```

```
WithdrawAsset (KLT, 100)
ReserveAssetDeposited(WETH, 2)
PayFees (WETH, 1)

> SetAssetClaimer Ethereum/Alice   // derived from beneficiary

Reviewer: Yes, this would be derived by the msg.sender on Ethereum. If the user interacts with the bridge indirectly through a smart contract than it will be that contracts address. If the user calls the bridge directly, through a dapp, then it will be the user address.

AliasOrigin Ethereum/Alice       // derived from msg.sender
InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT

> 		DepositAsset (KLT, 100) dest=Bob

Reviewer 2: ```suggestion
		DepositAsset (KLT, 100) beneficiary=Bob
```

		Transact Bob.hello()
```

## Step 4: KILT Receives XCM from AH

> The flowing XCM $x_1$ is received from BH on AH.

Reviewer 2: ```suggestion
The following XCM $x_1$ is received from BH on AH.
```

```
*WithdrawAsset (KLT, 100)
*ReserveAssetDeposited (WETH, 1)
*PayFees (WETH, 1)

> *SetAssetClaimer Bob
> *AliasOrigin Ethereum/Alice   // origin preserved from AH
> DepositAsset (KLT, 100) dest=Bob

Reviewer 2: ```suggestion
*SetAssetClaimer Ethereum/Alice
*AliasOrigin Ethereum/Alice   // origin preserved from AH
SetAssetClaimer Bob
DepositAsset (KLT, 100) beneficiary=Bob
```

```

```
*WithdrawAsset (KLT, 100)
*ReserveAssetDeposited (WETH, 1)
*PayFees (WETH, 1)

> *SetAssetClaimer Bob
> *AliasOrigin Ethereum/Alice   // origin preserved from AH
> DepositAsset (KLT, 100) dest=Bob

Reviewer 2: ```suggestion
*SetAssetClaimer Ethereum/Alice
*AliasOrigin Ethereum/Alice   // origin preserved from AH
SetAssetClaimer Bob
DepositAsset (KLT, 100) beneficiary=Bob
```

Transact Bob.hello()          // executes with the origin from AH
```

# Relayer Rewards

The tracking and disbursement of relayer rewards for both directions has been unified. Rewards in WETH are accumulated on BH but must be manually claimed to deposit them into an account.

To claim, call following extrinsic, where $o$ is rewards account (origin), and $w$ is account on AH where the WETH will be paid out.

$$
\mathrm{claim}(o,w)
$$

> For tax accounting purposes might be desirable that $o \neq w$.

Reviewer 2: ```suggestion
For tax accounting purposes it might be desirable that $o \neq w$.
```

# Top-Up

> Top-up on the destination is viable to implement.

Reviewer 2: ```suggestion
Top-up of the relayer reward is viable to implement for either direction as extrinsics on Bridge Hub, respectively Ethereum.
```

1. `AliasOrigin` can behave like `DescendOrigin`. This is allowed because it respects the hierarchy of multi-locations and does not allow jumping up.
    1. Example location `Ethereum` can alias into `Ethereum/Alice` because we are descending in origin and this essentially is how the `DescendOrigin` instruction works.
2. `AliasOrigin` must allow AH to alias into bridged locations such as `{ parents: 2, interior: GlobalConsensus(Ethereum) }` so that AH can act as a proxy for the bridge on parachains.

`AliasOrigin` will be inserted by every `InitiateAssetTransfer` instruction on the source parachain, populated with the contents of the origin register, essentially forwarding the origin of the source to the destination.

RFCS:

[https://github.com/polkadot-fellows/RFCs/pull/122](https://github.com/polkadot-fellows/RFCs/pull/122)

[https://github.com/polkadot-fellows/RFCs/blob/main/text/0100-xcm-multi-type-asset-transfer.md](https://github.com/polkadot-fellows/RFCs/blob/main/text/0100-xcm-multi-type-asset-transfer.md)

## Parachain Requirements

1. Pallet-xcm.execute enabled.
2. XCM payment dry run apis.

> 3. Must accept WETH needed for fees.

Reviewer 2: You could add a comment here that this is only temporarily mandatory, user agents can be improved in the future to inject `ExchangeAsset give=(WETH, 1) want=(KLT, 20)` in the message executing on AH.

- Off-chain fee estimation
- P→E Fee Asset: WETH
- E→P Fee Asset: ETH
- Relayer rewards for both directions paid out on AH in WETH

# Polkadot→Ethereum

Given source parachain $S$, with native token $S^{'}$ and the initial xcm $x_0$ to be executed on $S$.

## Step 1: User agent constructs initial XCM

The user agent constructs an initial XCM message $x_0$ that will be executed on S.

The fee amounts in this message should be high enough to enable dry-running, after which they will be lowered.

## Step 2: User agent estimates fees

- Given source parachain $S$, with native token $S^{'}$ and the initial xcm $x_0$ to be executed on $S$.
- The native currency $P^{'}$ of the Polkadot relay chain, and $E^{'}$ of Ethereum.
- Suppose that the user agent chooses relayer reward $r$ in $E^{'}$.
- Suppose that the exchange rates  $K_{P^{'}/S^{'}}$ and $K_{E^{'}/S^{'}}$. The user agent chooses a multiplier to $\beta$ to cover volatility in these rates.

Apply the following sequence operations:

1. Dry-run $x_0$ on $S$ to receive xcm $x_1$ and cost $a$ in $S^{'}$
2. Dry-run $x_1$ on AH to receive xcm $x_2$ and cost $b$ in DOT
3. Dry-run $x_2$ on BH to receive command $m$ and cost $c$ in DOT

> 4. Dry-run $m$ on Ethereum to receive cost $d$ in ETH

Author: The mathematical notation was useful for distilling the essence of the algorithm. However I've made it more clear in 495ab69b2f352a1f8af05b01dc36bd4df811eadc that DOT and ETH are alternative names.

The final cost to the user in $S^{'}$ is given by

$$
\beta \left(a + \frac{b + c}{K_{P^{'}/S^{'}}} + \frac{d + r}{K_{E^{'}/S^{'}}}\right)
$$

The user agent should perform a final update to xcm $x_0$, substituting the calculated fee amounts.

## Step 3: User agent initiates bridging operation

The user agent calls `pallet_xcm::execute` with the initial xcm $x_0$

```
WithdrawAsset (KLT, 100)
PayFees (KLT, 20)
InitiateAssetsTransfer asset=(KLT, 60) remoteFee=(KLT, 20) dest=AH
	ExchangeAsset give=(KLT, 20) want=(WETH, 1)
	InitiateAssetsTransfer asset=(KLT, 40) remoteFee=(WETH, 1) dest=Ethereum
		DepositAsset (KLT, 40) beneficiary=Bob
```

## Step 4:  AH executes message x1

The message $x_1$ is application-specific:

```
ReserveAssetDeposited (KLT, 80)
PayFees (KLT, 20)
SetAssetClaimer Kilt/Alice
AliasOrigin Kilt/Alice
ExchangeAsset give=(KLT, 20) want=(WETH, 1)
InitiateAssetsTransfer asset=(KLT, 60) remoteFee=(WETH, 1) dest=Ethereum
	DepositAsset (KLT, 60) beneficiary=Bob
```
or
```
*ReserveAssetDeposited (KLT, 80)
*PayFees (KLT, 20)

> *SetAssetClaimer Kilt/Alice

Reviewer 3: I was confused at first with this, since after `AliasOrigin`, the claimer of the assets will be `Kilt/Alice`. However, I realized that `AliasOrigin` itself might fail, if the chain is not configured to accept that. In that case it makes sense to do `SetAssetClaimer` before with the same exact origin.

A little note would be helpful to others like me that got might get confused about it 😅

1. Pay fees for local execution using `PaysFees`
2. Obtain WETH for remote delivery fees.

The XCM bridge-router on AH will charge a small fee to prevent spamming BH with bridge messages. This is necessary since the `ExportMessage` instruction in message $x_2$ will have no execution fee on BH. For a similar reason, we should also impose a minimum relayer reward of at least the existential deposit 0.1 DOT, which acts as a deposit to stop spamming messages with 0 rewards.

## Step 5: BH executes message x2

Message $x_2$ is parsed by the `SnowbridgeMessageExporter` with the following effects:

- A bridge command $m$ is committed to MMR structure $M$.
    - The transferred asset is parsed from `ReserveAssetDeposited` , `WithdrawAsset` or `TeleportedAssetReceived` instructions for the local, destination and teleport asset transfer types respectively.
    - The original origin is preserved through the `AliasOrigin` instruction. This will allow us to resolve agents for the case of `Transact`.
    - The message exporter must be able to support multiple assets and reserve types in the same message and potentially multiple `Transacts`.

>     - The Message Exporter must be able to support multiple Deposited Assets.

Reviewer: Correct. It is the `SnowbridgeExporter` on Bridge Hub.

    - The Message Exporter must be able to parse `SetAssetClaimer` and allow the provided location to claim the assets on BH in case of errors.

> - Given relayer reward $r$, set storage $P(\mathrm{hash}(m)) = r$. This is parsed from the `WithdrawAsset` and `PayFees` instruction.

Author: Clarified in 1c0c9f0481034ccbe1d06b9804f8fd894f270467

	*ReserveAssetDeposited (KLT, 60)
	*WithdrawAsset (WETH, 1)
	*PayFees (WETH, 1)
	*SetAssetClaimer Kilt/Alice
	*AliasOrigin Kilt/Alice
	DepositAsset (KLT, 60) beneficiary=Bob
```
or
```
!WithdrawAsset(DOT, 10)
!PayFees (DOT, 10)
!ExportMessage dest=Ethereum
	*ReserveAssetDeposited (KLT, 80)
	*PayFees (KLT, 20)
	*SetAssetClaimer Kilt/Alice
	*AliasOrigin Kilt/Alice
	DepositAsset (KLT, 60) beneficiary=Bob
	Transact Bob.hello()
```

## Step 6: Relayer relays message to Gateway

1. A relayer *Charlie* inspects storage $P$ to look for new messages to relay. Suppose it finds $\mathrm{hash}(m)$ giving reward $r$.
2. The relayer queries $m$ from $M$ and constructs the necessary proofs.
3. The relayer dry-runs m on Ethereum to decide whether the message is profitable to deliver.
4. The relayer finally delivers the message together with a relayer-controlled address $u$ on AH where the relayer reward will be lazily deposited.

## Step 7: Relayer delivers proof of delivery to BH

The proof of delivery is essentially a merkle proof for the `InboundMessageAccepted` event log.

When BH processes the proof of delivery:

1. The command $m$ is removed from storage items $M$ and $P$.

> 2. The relayer reward is tracked in storage $R$, where $R(u)$ is the accumulated rewards that can be claimed by account $u$.

Author: I've made this more clear in the other section where we expand on this dc7d01a83a515b4e99545d15b8f068e696788ea9

3. The same interface is used for both PNA (Polkadot Assets) and ERC20 tokens. Internally we will still look up whether the token is registered as a PNA or ERC20 for the purpose of minting/locking burning/unlocking logic. The asset transfer type chosen by the UX layer will inform the XCM that is built from the message on BH.

```solidity
enum Kind {
    Index,
    Address32,
    Address20,
    XCMPayload,
}

struct Beneficiary {
    Kind kind;
    bytes data;
}

enum AssetTransferType {
	ReserveDeposit, ReserveWithdraw, Teleport
}

struct Token {
	AssetTransferType type;
	address token;
	uint128 amount;
}

function send(
  ParaID destinationChain,
  Beneficiary calldata beneficiary,
	Token[] tokens,
  uint128 reward) external payable;
```

Message enqueued $m_0$:

```solidity
send(
	3022, // KILT Para Id
	Address32(0x0000....),
	[
		(ReserveWithdraw, KLT, 100)
	],
	10, // WETH
	)
```

```solidity
send { value: 3 }( // Send 3 Eth for fees and reward
	3022, // KILT Para Id
	XCMPayload(
		DepositAsset (KLT, 100) dest=Bob
		Transact Bob.hello()
	),
	[
		(ReserveWithdraw, KLT, 100)
	],
	1, // 1 ETH of 3 needs to be for the reward, the rest is for fees
	)
```

> The User Agent/UX layer will need to estimate the fee required to be passed into the `send` method. This may be an issue as we cannot Dry-Run something on Polkadot that has not even been submitted on Ethereum yet. We may need to make RPC API to DryRun and get back the xcm that would be submitted to asset hub.

Reviewer 3: Why do we need to submit it on Ethereum first? Can't we dry-run the XCM that happens as a result of this command `m`?

Author: I think we will need a helper API in our gateway contract that can be used in an off-chain context to produce the  messages/commands that can be used by the dry-run API on BH.

The message is converted from $m_0$ to $x_0$ during message submission on BH. Dry-running submission will return $x_0$ to the relayer so that it can verify it is profitable.

## Step 3: AH receives $x_0$ from BH

Submitting the message $m_0$ will cause the following XCM, $x_0$, to be built on BH and dispatched to AH.

```
WithdrawAsset (KLT, 100)
ReserveAssetDeposited(WETH, 2)
PayFees (WETH, 1)
SetAssetClaimer Bob            // derived from beneficiary
AliasOrigin Ethereum/Alice     // derived from msg.sender
InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
	DepositAsset (KLT, 100) beneficiary=Bob
```

```
WithdrawAsset (KLT, 100)
ReserveAssetDeposited(WETH, 2)
PayFees (WETH, 1)
SetAssetClaimer Ethereum/Alice   // derived from beneficiary
AliasOrigin Ethereum/Alice       // derived from msg.sender
InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
		DepositAsset (KLT, 100) beneficiary=Bob
		Transact Bob.hello()
```

## Step 4: KILT Receives XCM from AH

> The following XCM $x_1$ is received from BH on AH.

Reviewer: ```suggestion
The following XCM $x_1$ is received from AH on KILT.
```

```
*WithdrawAsset (KLT, 100)
*ReserveAssetDeposited (WETH, 1)
*PayFees (WETH, 1)
*SetAssetClaimer Ethereum/Alice
*AliasOrigin Ethereum/Alice   // origin preserved from AH
SetAssetClaimer Bob

> DepositAsset (KLT, 100) beneficiary=Bob

Reviewer 3: This should be a teleport if we want another message to go to Kilt

Reviewer: This is the XCM that KILT will receive from AH. And it assumes reserve-based transfer. If in reality KILT does teleports then `InitiateAssetsTransfer` should take care of doing the right thing here provided we specify the correct reserve type on Asset Hub.

```

```
*WithdrawAsset (KLT, 100)
*ReserveAssetDeposited (WETH, 1)
*PayFees (WETH, 1)
*SetAssetClaimer Ethereum/Alice
*AliasOrigin Ethereum/Alice   // origin preserved from AH
SetAssetClaimer Bob
DepositAsset (KLT, 100) beneficiary=Bob

> Transact Bob.hello()          // executes with the origin from AH

Reviewer 3: Same

```

# Relayer Rewards

The tracking and disbursement of relayer rewards for both directions has been unified. Rewards in WETH are accumulated on BH but must be manually claimed to deposit them into an account.

> To claim, call following extrinsic, where $o$ is rewards account (origin), and $w$ is account on AH where the WETH will be paid out.

Reviewer 3: This `o` has to be the same `u` as described previously, right? That way we know they're able to claim these rewards. I'd add that here if that's the case.

$$
\mathrm{claim}(o,w)
$$

For tax accounting purposes it might be desirable that $o \neq w$.

# Top-Up

> Top-up of the relayer reward is viable to implement for either direction as extrinsics on Bridge Hub, respectively Ethereum.

Reviewer 3: ```suggestion
Top-up of the relayer reward is viable to implement for either direction as extrinsics on Bridge Hub and Ethereum respectively.
```
?

Didn't understand the original phrase.

## Origin Preservation

Origins for transact will be preserved by use of the `AliasOrigin` instruction. This instruction will have the following rules that parachain runtimes will need to allow:

> 1. `AliasOrigin` can behave like `DescendOrigin`. This is allowed because it respects the hierarchy of multi-locations and does not allow jumping up.

Reviewer 3: ```suggestion
1. `AliasOrigin` can behave like `DescendOrigin`. This is safe because it respects the hierarchy of multi-locations and does not allow jumping up. Meaning no escalation of privileges.
```

    1. Example location `Ethereum` can alias into `Ethereum/Alice` because we are descending in origin and this essentially is how the `DescendOrigin` instruction works.

> 2. `AliasOrigin` must allow AH to alias into bridged locations such as `{ parents: 2, interior: GlobalConsensus(Ethereum) }` so that AH can act as a proxy for the bridge on parachains.

Reviewer 3: ```suggestion
2. `AliasOrigin` must allow AH to alias into bridged locations such as `{ parents: 2, interior: GlobalConsensus(Ethereum) }` and all of its internal locations so that AH can act as a proxy for the bridge on parachains.
```

> 2. `AliasOrigin` must allow AH to alias into bridged locations such as `{ parents: 2, interior: GlobalConsensus(Ethereum) }` so that AH can act as a proxy for the bridge on parachains.

Reviewer 2: No need to allow AH to "alias anything". The following is enough:

on AH:
- Allow any location to alias into child locations,

on Paras:
- Allow AH to alias into bridged locations.

Aliasing `Kilt/Bob` happens on AH for message coming from `Kilt` (but initiated by `Bob` on `Kilt`).

`AliasOrigin` will be inserted by every `InitiateAssetTransfer` instruction on the source parachain, populated with the contents of the origin register, essentially forwarding the origin of the source to the destination.

RFCS:

[https://github.com/polkadot-fellows/RFCs/pull/122](https://github.com/polkadot-fellows/RFCs/pull/122)

[https://github.com/polkadot-fellows/RFCs/blob/main/text/0100-xcm-multi-type-asset-transfer.md](https://github.com/polkadot-fellows/RFCs/blob/main/text/0100-xcm-multi-type-asset-transfer.md)

## Parachain Requirements

1. Pallet-xcm.execute enabled.

> 2. XCM payment dry run apis.

Reviewer 3: ```suggestion
2. XCM payment and dry run apis implemented.
```

3. Must accept WETH needed for fees.

> 4. AH as a reserve for bridged assets.

Reviewer 3: ```suggestion
4. Trust AH as a reserve for bridged assets.
```
